File: jenkins-plugin/src/main/java/org/eazyportal/plugin/release/jenkins/project/ProjectDescriptorFactory.java
package org.eazyportal.plugin.release.jenkins.project;

import hudson.Extension;
import hudson.FilePath;
import hudson.model.InvisibleAction;
import org.eazyportal.plugin.release.core.project.model.ProjectDescriptor;
import org.eazyportal.plugin.release.core.project.model.ProjectFile;
import org.eazyportal.plugin.release.core.scm.ScmActions;
import org.eazyportal.plugin.release.gradle.project.GradleProjectActionsFactory;

import javax.inject.Inject;
import java.io.Serializable;

@Extension
public class ProjectDescriptorFactory extends InvisibleAction implements Serializable {

    @Inject
    private transient GradleProjectActionsFactory projectActionsFactory;

    private final transient org.eazyportal.plugin.release.core.project.ProjectDescriptorFactory<FilePath> delegate =
        new org.eazyportal.plugin.release.core.project.ProjectDescriptorFactory<>();

    public ProjectDescriptor<FilePath> create(ProjectFile<FilePath> projectFile, ScmActions<FilePath> scmActions) {
        return delegate.create(projectActionsFactory, scmActions, projectFile);
    }

}

File: core/src/main/kotlin/org/eazyportal/plugin/release/core/project/ProjectDescriptorFactory.kt
package org.eazyportal.plugin.release.core.project

import org.eazyportal.plugin.release.core.project.model.Project
import org.eazyportal.plugin.release.core.project.model.ProjectDescriptor
import org.eazyportal.plugin.release.core.project.model.ProjectFile
import org.eazyportal.plugin.release.core.scm.ScmActions

open class ProjectDescriptorFactory<T> {

    fun create(
        projectActionsFactory: ProjectActionsFactory,
        scmActions: ScmActions<T>,
        rootProjectFile: ProjectFile<T>
    ): ProjectDescriptor<T> {
        val rootProject: Project<T> = rootProjectFile.toProject(projectActionsFactory)

        val subProjects: List<Project<T>> = scmActions.getSubmodules(rootProject.dir)
            .map { rootProject.dir.resolve(it) }
            .map { it.toProject(projectActionsFactory) }

        return ProjectDescriptor(
            allProjects = listOf(*subProjects.toTypedArray(), rootProject),
            rootProject = rootProject,
            subProjects = subProjects
        )
    }

    private fun ProjectFile<T>.toProject(projectActionsFactory: ProjectActionsFactory): Project<T> =
        Project(this, projectActionsFactory.create(this))

}
